Assessments
===========

People come to this class with very different previous experiences. Please take
each of the following tests so that we know have a general idea of your
history. It doesn't matter how high you score on any of these assessments.

## Molecular Biology ##

Give yourself 1 point for each molecular biology concept you can explain.

+ alpha helix
+ amino acid
+ beta sheet
+ C0t curve
+ CDS
+ chromosome
+ codon
+ consensus sequence
+ DNA
+ frame-shift
+ genome
+ enhancer
+ exon
+ hybridization
+ interspersed repeat
+ intron
+ inversion
+ kinase
+ Kozak consensus
+ motif
+ mRNA
+ oligo
+ operon
+ ORF
+ nucleic acid
+ PCR
+ plasma membrane
+ poly-A signal
+ promoter
+ protein
+ reading frame
+ replication
+ restriction digest
+ restriction site
+ reverse complement
+ ribosome
+ RNA-seq
+ Shine-Dalgarno sequence
+ shotgun sequencing
+ signal peptide
+ SNP
+ splice acceptor site
+ splice donor site
+ start codon
+ stop codon
+ TATA box
+ transcription
+ translation
+ translocation
+ vector

## Programming ##

Give yourself 1 point for each programming concept you can explain.

+ abstraction
+ assertion
+ binary notation
+ binary search
+ Boolean operators
+ buffer
+ casting
+ closure
+ compiler
+ const
+ doc-strings
+ dynamic linkage
+ dynamic typing
+ filo
+ floating point limits
+ function protoype
+ garbage collection
+ git
+ hash function
+ heap
+ hexadecimal notation
+ in-lining
+ inner join
+ isa
+ integer types
+ library
+ make
+ malloc
+ man page
+ namespace
+ parent class
+ pointer arithmetic
+ pragma
+ private
+ recursion
+ reference
+ reference count
+ return value
+ shared object file
+ stack
+ static
+ struct
+ swap memory
+ thread-safe
+ typedef
+ unit test
+ unsigned
+ usage statement
+ variable interpolation
+ vector

## Bioinformatics ##

The following terms are a mixture of technologies that are often associated
with bioinformatics. Give yourself 1 point for each item you can explain.

+ amino acid single letters
+ anaconda
+ Bayesian statistics
+ BLAST
+ BLOSUM matrix
+ bowtie
+ containerization
+ dynamic programming
+ expectation maximization
+ FASTA
+ FASTQ
+ frequentist statistics
+ Galaxy
+ GFF
+ Gibbs sampling
+ GitHub
+ GO (not the language)
+ ggplot
+ HMM
+ indel
+ JSON
+ Karlin-Altschul statistics
+ K-mer
+ log-odds ratio
+ motif
+ multiple alignment
+ Needleman-Wunsch
+ nucleotide ambiguity codes
+ pairwise alignment
+ PAM matrix
+ pandas
+ peak-calling
+ Phred
+ pipeline
+ Profile HMM
+ PWM
+ R
+ rsync
+ SAM file
+ slurm
+ Smith-Waterman
+ snakemake
+ SQL
+ tensorflow
+ Viterbi
+ virtual environment
+ virtual machine
+ virtual memory
+ YAML
+ Z-score

